libs/wire-subsystems/src/Wire/EnterpriseLoginSubsystem/Interpreter.hs

```diff
@@ -0,0 +1,321 @@
+  upsert $ toStored new
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/lock"
+  sendAuditMail url "Domain locked" mOld (Just new)
+
+unlockDomainImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+unlockDomainImpl domain = do
+  old <- getDomainRegistrationImpl domain
+  let new = old {domainRedirect = None} :: DomainRegistration
+  case old.domainRedirect of
+    Locked -> upsert $ toStored new
+    _ -> throw EnterpriseLoginSubsystemUnlockError
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/unlock"
+  sendAuditMail url "Domain locked" (Just old) (Just new)
+
+preAuthorizeImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+preAuthorizeImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  new <- case mOld of
+    Nothing -> do
+      let new = DomainRegistration domain PreAuthorized Allowed Nothing
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == None -> do
+      let new = sdr {domainRedirect = PreAuthorized} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == PreAuthorized -> pure sdr
+    _ -> throw EnterpriseLoginSubsystemErrorInvalidDomainRedirect
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/preauthorize"
+  sendAuditMail url "Domain locked" mOld (Just new)
+
+getDomainRegistrationImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r
+  ) =>
+  Domain ->
+  Sem r DomainRegistration
+getDomainRegistrationImpl domain = do
+  mSdr <- tryGetDomainRegistrationImpl domain
+  case mSdr of
+    Just dr -> pure dr
+    Nothing -> throw EnterpriseLoginSubsystemErrorNotFound
+
+tryGetDomainRegistrationImpl ::
+  forall r.
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r
+  ) =>
+  Domain ->
+  Sem r (Maybe DomainRegistration)
+tryGetDomainRegistrationImpl domain = do
+  mSdr <- lookup domain
+  maybe (pure Nothing) (fmap Just . deserialize) mSdr
+  where
+    deserialize :: StoredDomainRegistration -> Sem r DomainRegistration
+    deserialize sdr = do
+      let mDomainRegistration = fromStored sdr
+      case mDomainRegistration of
+        Nothing -> do
+          Log.err $ Log.field "domain" (toByteString' domain) . Log.msg (Log.val "Invalid stored domain registration")
+          throw EnterpriseLoginSubsystemInternalError
```

Comment:
given this is an internal end-point, maybe we can add more info to the error response?  starting with `sdr` in full?